src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Provides a high-performance API for forward-only, read-only access to the UTF-8 encoded JSON text.
+    /// It processes the text sequentially with no caching and adheres strictly to the JSON RFC
+    /// by default (https://tools.ietf.org/html/rfc8259). When it encounters invalid JSON, it throws
+    /// a JsonReaderException with basic error information like line number and byte position on the line.
+    /// Since this type is a ref struct, it does not directly support async. However, it does provide
+    /// support for reentrancy to read incomplete data, and continue reading once more data is presented.
+    /// To be able to set max depth while reading OR allow skipping comments, create an instance of 
+    /// <see cref="JsonReaderState"/> and pass that in to the reader.
+    /// </summary>
+    public ref partial struct JsonUtf8Reader
+    {
+        private ReadOnlySpan<byte> _buffer;
+
+        private bool _isFinalBlock;
+
+        private ulong _stackFreeContainer;
+        private long _lineNumber;
+        private long _lineBytePosition;
+        private int _consumed;
+        private int _currentDepth;
+        private int _maxDepth;
+        private bool _inObject;
+        private bool _isNotPrimitive;
+        private JsonTokenType _tokenType;
+        private JsonReaderOptions _readerOptions;
+        private Stack<JsonTokenType> _stack;
```

Comment:
Is there a perf benefit to storing all of these individually rather than just having a single JsonReaderState field and having all accesses be to members of that field?  Then the ctor would just store the state in one assignment, and the CurrentState property would just return the field.

Reply:
> Is there a perf benefit to storing all of these individually rather than just having a single JsonReaderState field and having all accesses be to members of that field? Then the ctor would just store the state in one assignment, and the CurrentState property would just return the field.

I thought about doing it, but haven't measured the impact. The reason is that the `JsonReaderState` contains two fields that I want to ignore when creating the reader but keep on the state as its relevant to the caller.
- BytesConsumed
- Position

I do not honor the `BytesConsumed` coming from the state when I create a new instance of the `JsonUtf8Reader`. I retet the private `_consumed` field to 0. So, if I had a state field instead, I would end up growing the struct by having an unnecessary copy of the field (one from the state struct, and one of my own). But maybe that's fine.

> On that note, did you explore just having the JsonReaderState passed into the ctor by ref? 

Yes I did explore it and I plan to continue to investigating the viability of that approach. If successful, I will make that change.

Reply:
TODO: https://github.com/dotnet/corefx/issues/33298